src/Std/Time/Duration.lean

```diff
@@ -0,0 +1,132 @@
+/-
+Copyright (c) 2024 Lean FRO, LLC. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Sofia Rodrigues
+-/
+prelude
+import Std.Time.DateTime
+import Std.Time.Zoned
+
+namespace Std
+namespace Time
+open Internal
+
+set_option linter.all true
+
+/--
+Duration is just a period between two timestamps.
+-/
+def Duration := Timestamp
+  deriving Repr, BEq
+
+instance : ToString Duration where
+  toString s :=
+    let (sign, secs, nanos) :=
+      if s.second.val > 0 then ("" ,s.second.val, s.nano.val)
+      else if s.second.val < 0 then ("-", -s.second.val, -s.nano.val)
+      else if s.nano.val < 0 then ("-", -s.second.val, -s.nano.val) else ("", s.second.val, s.nano.val)
+    sign ++ toString secs ++ "." ++ toString nanos ++ "s"
+
+instance : Repr Duration where
+  reprPrec s := reprPrec (toString s)
+
+namespace Duration
+
+/--
+Converts a `Duration` to a `Minute.Offset`
+-/
+def toMinutes (tm : Duration) : Minute.Offset :=
+  tm.second.ediv 60
+
+/--
+Converts a `Duration` to a `Day.Offset`
+-/
+def toDays (tm : Duration) : Day.Offset :=
+  tm.second.ediv 86400
+
+/--
+Creates a new `Duration` out of `Second.Offset`.
+-/
+def ofSeconds (secs : Second.Offset) : Duration :=
+  Timestamp.ofSecondsSinceUnixEpoch secs
+
+/--
+Creates a new `Duration` out of `Nanosecond.Offset`.
+-/
+def ofNanosecond (secs : Nanosecond.Offset) : Duration :=
+  Timestamp.ofNanosecondsSinceUnixEpoch secs
+
+/--
+Calculates a `Duration` out of two `Timestamp`s.
+-/
+def since (f : Timestamp) : IO Duration := do
+  let cur ← Timestamp.now
+  return Std.Time.Timestamp.sub cur f
+
+/--
+Adds a `Duration` to a `Timestamp`.
+-/
+def add (f : Timestamp) (d : Duration) : Timestamp :=
+  Std.Time.Timestamp.add f d
+
+/--
+Checks if the duration is zero seconds and zero nanoseconds.
+-/
+def isZero (d : Duration) : Bool :=
+  d.second.val = 0 ∧ d.nano.val = 0
+
+instance : HAdd Timestamp Duration Timestamp where
+  hAdd := add
+
+instance : HSub Timestamp Timestamp Duration where
+  hSub := Std.Time.Timestamp.sub
+
+end Duration
+
+namespace PlainDate
+
+/--
+Calculates the duration between a given `PlainDate` and a specified date.
+-/
+def since [ToTimestamp α] (date : PlainDate) (since : α) : Duration :=
+  let date  := date.toTimestamp
+  let since := ToTimestamp.toTimestamp since
+  Std.Time.Timestamp.sub date since
+
+end PlainDate
+
+namespace PlainDateTime
+
+/--
+Calculates the duration between a given `PlainDateTime` and a specified date.
+-/
+def since [ToTimestamp α] (date : PlainDateTime) (since : α) : Duration :=
```

Comment:
Something is clearly wrong here!
```lean
import Std.Time

open Std.Time

def now := PlainDateTime.ofUTCTimestamp ⟨1724859638, ⟨907328169, by decide⟩, by decide⟩
def after := PlainDateTime.ofUTCTimestamp ⟨1724859639, ⟨907641224, by decide⟩, by decide⟩
def duration := after.since now
#eval now.format "HH:mm:ss.ssssssss" -- "03:40:38.00000038"
#eval after.format "HH.mm:ss.ssssssss" -- "03.40:39.00000039"
#eval duration -- "1.313055s"
```

Reply:
The problem was in the pretty printing of the duration and `ssssssss` that needs 9 s to format nanoseconds. Fixed it and added a test.